### ISSUE TITLE
Forward Capabilities Outside Admin

### ISSUE THREAD

**Issue by katelynsills** [yourlabs/django-autocomplete-light#756]
I'm currently trying to use the forward widget argument in a non-admin context, and it looks as though the forward.js script expects a different html element hierarchy than what I have. Please let me know if I should post this elsewhere, or provide more of a minimal test case in order to reproduce it, but here's my current implementation. 

There is a "Box" model with the following fields, where the choices for the raw_product should depend on the status choice.

```
class Box(models.Model):
    raw_product = models.ForeignKey(RawProduct)
    status = models.CharField(max_length=20, choices=STATUS_CHOICES, null=True)
```

I've created a form to test the forward capabilities:

```
class DalBoxForm(ModelForm):
    status = ChoiceField(choices=STATUS_CHOICES)

    class Meta:
        model = Box
        fields = ('__all__')
        widgets = {
            'raw_product': autocomplete.ModelSelect2(url='raw-product-status-autocomplete', forward=['status'])
        }
```

And the following view: 

```
class RawProductStatusAutocomplete(autocomplete.Select2QuerySetView):
    def get_queryset(self):
        if not self.request.user.is_authenticated():
            return RawProduct.objects.none()

        qs = RawProduct.objects.all()
        status = self.forwarded.get('status', None)

        if status:
            qs = qs.filter(short='YPC-B') # doesn't make logical sense, but this is to test that it is working only

        if self.q:
            qs = qs.filter(long__istartswith=self.q)

        return qs
```

It looks like select2.js gets the element identified with "[data-autocomplete-light-function=select2]" and forward.js looks for siblings of that element that are of class "dal-forward-conf". However, in the above example, the element with class "dal-forward-conf" is not a sibling of the element. 

Here's the html created by the form:

```
<p>
    <label for="id_raw_product">Raw product:</label>
    <select aria-hidden="true" class="select2-hidden-accessible"
            data-autocomplete-light-function="select2"
            data-autocomplete-light-url="/raw-product-status-autocomplete/"
            id="id_raw_product" name="raw_product" required="" tabindex="-1">
        <option value="HV">Hairy Vetch</option>
    </select>
    <span class="select2 select2-container select2-container--default select2-container--below select2-container--open"
          dir="ltr" style="width: 808px;">
        <span class="selection">
            <span aria-expanded="true" aria-haspopup="true"
               aria-labelledby="select2-id_raw_product-container"
               aria-owns="select2-id_raw_product-results"
               class="select2-selection select2-selection--single"
               role="combobox" tabindex="0">
                <span class="select2-selection__rendered" id= "select2-id_raw_product-container" title="Hairy Vetch">
                    <span class="select2-selection__clear">×</span>
                    Hairy Vetch
                </span>
                <span class="select2-selection__arrow" role="presentation">
                    <b role="presentation"></b>
                </span>
            </span>
        </span>
        <span aria-hidden="true" class="dropdown-wrapper"></span>
    </span>
</p>
<div style="display:none" class="dal-forward-conf" id="dal-forward-conf-for-id_raw_product">
    <script type="text/dal-forward-conf">[{"src": "status", "type": "field"}]</script>
</div>
```

To get the forwarded element, you'd have to do something like:
```
var element = $('[data-autocomplete-light-function=select2]')
element.parent().siblings("div.dal-forward-conf")
```

I'd be happy to submit a pull request, but I'm not sure if this is a common issue or something pertaining to my unique setup.

**Comment by jpic** [comment]
> the choices for the raw_product should depend on the status choice.

Do you mean _must_ instead of _should_ ? You should know that the forward feature does not offer any kind of security whatsoever during input validation. For this, you'd have to find a way like [ddf](https://github.com/yourlabs/django-dynamic-fields).

> To get the forwarded element, you'd have to do something like

Perhaps you can override the javascript method ? Otherwise perhaps try the common ancestor like [in ddf](https://github.com/yourlabs/django-dynamic-fields/blob/master/src/ddf/static/ddf/ddf.js#L175-L179) ?

> I'd be happy to submit a pull request, but I'm not sure if this is a common issue or something pertaining to my unique setup.

It is tested against the django.contrib.admin vanilla templates only. I wouldn't be suprised some code could support more configurations with nothing but small fixes.

Another option I'd like to compare with that is an implementation of `dal/forward.js` in ddf. Leveraging the Action(Conditions) based architecture in ddf. These small units also permit much easier JS TDD. Another advantage is that server side security for this feature definitely has nothing to do in dal, so the place we could maintain it is in an app like ddf.

Ref #732

**Comment by katelynsills** [comment]
Ah, ok, I didn't realize that this feature wasn't expected to work in a non-admin context. I'm not very familiar with the inner workings of django autocomplete light, but this would definitely be a feature I'd be interested in.

I'm not sure what you mean about the distinction between _should_ and _must_. (I was trying to provide an example where the forward.js script wasn't working as I was expecting, so the model fields in the example don't correspond exactly to real business logic.) DDF definitely looks like it may solve this issue, though, so I'll take a further look there. Thank you for the recommendation!

**Comment by jpic** [comment]
> I'm not sure what you mean about the distinction between _should_ and _must_.

If the choice of the second field *must* depend on the choice of the first field, then you need server side validation too, which dal doesn't provide, but that you could add to your form's clean() method.

> Ah, ok, I didn't realize that this feature wasn't expected to work in a non-admin context

Well, it's not _not expected_ per say, if you use the same HTML rendering function than the admin, which is probably something like `{{ form.as_table }}` then it would probably work out of the box. It's not that it shouldn't support any other DOM trees than the one in the admin, just it was developed with the admin for reference. 

If we really want to be able to support all sorts of DOM trees then the best is to just let users plug their callback. If you can reproduce this in a new app outside the admin in the test_project that would help moving on forward. 

> this would definitely be a feature I'd be interested in.

I'd tend to think that it would be worth changing the couple lines of code that your use cas needs in dal, if you just want to get this working and move on it would definitely work. So, that's fine, we can make it work. 

But if you want more than something that works in the short run, if you're not just here to get a project done but you find it more fun and profitable to invest in long term to research the best way that you're going to reuse for as long as you're going to enjoy hacking with django, then ddf would of course seem to be the right direction, even though I can't promise nothing better is going to come out from somewhere else in the world tomorrow ^^ Not to mention that it'll be a lot more expensive and time consuming in the short run - so you'll probably need to have a fix for the current situation anyway.

PRs are ok on ddf as is, although there's something else I'd like to do before going on adding new features. I'm currently busy putting together a nice gulpfile to facilitate the hacking workflow and code in ES6 rather than ES5 to enjoy [ES6 classic class definitions](http://es6-features.org/#ClassInheritance) which would work a lot better for something like this - having matching JS classes for Python classes. Also I'm trying to have it setup requirejs and browserify properly because I secretely hope to be able to leverage better class imports than what [i've put together so far](https://github.com/yourlabs/django-dynamic-fields/blob/master/src/ddf/static/ddf/ddf.js#L4-L45) - that was heavily inspired by @gagarski's work on the forward feature in dal.

**Comment by gagarski** [comment]
@jpic The reason why it is broken is the my new format for forward configuration (inside `<script>` instead of HTML attribute). It was previously working outside Django admin templates and definitely was not supposed to be broken after updating to DAL 3.2. I'll try to look at it.

@katelynsills could you please provide a template which you use to render your form or (even better) a Django project with minimalistic broken example?

**Comment by katelynsills** [comment]
Thanks @gagarski. I can provide the Django project with a minimal example hopefully sometime this afternoon.

**Comment by jpic** [comment]
@gagarski this feature was released in 3.2 BTW, so perhaps another commit broke it ?

**Comment by gagarski** [comment]
@jpic Given the @katelynsills example, I don't think so (but I still want to play around with full examle). I render `<div class="dal-forward-conf>"<script> ... </script></div>` as a sibling to `<select>` element (while rendering a widget) and suppose to be a sibling after Select2 JS magic. Seems like my assumption is wrong.

In DAL 3.1 forward configuration was bound to `<select>` element with an HTML-attribute, so DOM-structure was not critical. 

Perhaps I should bound forward configuration to `<select>` more explicitly and look for it not only among `<select>` siblings but inside document or at least `<form>` tag.

**Comment by jpic** [comment]
@gagarski perhaps you could add the field name as a data- attribute of the script tag, then it could be anywhere in the DOM ? that's sort of how ddf works (to the difference that ddf outputs one script tag for all the form fields). Then, you don't need the script tag to be anywhere specific you could even render it from Media, similarly to what [django-content-editor](https://github.com/matthiask/django-content-editor/blob/master/content_editor/admin.py#L93) does.

**Comment by gagarski** [comment]
> perhaps you could add the field name as a data- attribute of the script tag, then it could be anywhere in the DOM ? 

I really considered that way while implementing advanced forward-conf. Unfortunately this apporach is not very friendly with admin inline dynamic forms (See Django's `admin/js/inline.js` and Ctrl-F `updateElementIndex`. This function updates only `for`, `id`, and `name` attributes of an inline form when copying it). The wrapping `<div>` exists thank to [django-dynamic-formset](https://github.com/elo80ka/django-dynamic-formset) (an attempt to bring inline forms outside of Django admin). 

So, `id="dal-forward-conf-for-field_name"` is a bit ugly but it works with admin inline forms and formsets and I do not see another way to deal with it now.

**Comment by gagarski** [comment]
@katelynsills if you need forward feature ASAP in your project, you can temporarily downgrade to DAL 3.1.8, your example should work properly with it.

**Comment by katelynsills** [comment]
@gagarski I think I have a working sample app in test_project locally. How would you prefer that I commit it? Should I create a new branch? For downgrading, I was using `Select2ListView` which I think is only available in 3.2.0, so it doesn't seem like I can downgrade easily, but that's ok.

**Comment by jpic** [comment]
The easiest for me is to just add an app to the existing test project with just the specific code that needs attention, then push it anywhere and link it here

**Comment by gagarski** [comment]
I attempted to fix it [here](https://github.com/yourlabs/django-autocomplete-light/commit/008d1a73f6e25a75c6877a99c09341cf24c5dab7), so if it works, i'll merge it into master.

**Comment by katelynsills** [comment]
@gagarski that fix works for all of my test cases

**Comment by gagarski** [comment]
Fixed in 82469ea2e63de0a4a96702abe842bba82e58ed0b
